src/units.cpp

```diff
@@ -464,6 +464,86 @@ double Units::scalingFactor(const UnitsPtr &units1, const UnitsPtr &units2)
     return 0.0;
 }
 
+using unitsMap = std::map<std::string, double>;
+
+void updateUnitsMap(const UnitsPtr &units, unitsMap &unitsMap, double exp = 1.0)
+{
+    if (units->isBaseUnit()) {
+        auto found = unitsMap.find(units->name());
+        if (found == unitsMap.end()) {
+            unitsMap.emplace(units->name(), 1.0);
+        } else {
+            found->second += exp;
+        }
+    } else {
+        for (size_t i = 0; i < units->unitCount(); ++i) {
+            std::string ref;
+            std::string pre;
+            std::string id;
+            double expMult;
+            units->unitAttributes(i, ref, pre, exp, expMult, id);
+            if (isStandardUnitName(ref)) {
+                auto unit = standardUnitsList.find(ref);
+                for (const auto &u : unit->second) {
+                    if (unitsMap.find(u.first) == unitsMap.end()) {
+                        unitsMap.emplace(u.first, u.second * exp);
+                    } else {
+                        auto ut = unitsMap.find(u.first);
+                        ut->second += u.second * exp;
+                    }
+                }
+            } else {
+                auto model = owningModel(units);
+                if (model != nullptr) {
+                    auto refUnits = model->units(ref);
+                    if ((refUnits == nullptr) || refUnits->isImport()) {
+                        unitsMap.clear();
+                        break;
+                    }
+                    updateUnitsMap(refUnits, unitsMap, exp);
+                }
+            }
+        }
+    }
+}
+
+bool Units::equivalent(const UnitsPtr &units1, const UnitsPtr &units2)
+{
+    // Initial checks.
+    if ((units1 == nullptr) || (units2 == nullptr)) {
+        return false;
+    }
+    if ((units1->isImport()) || (units2->isImport())) {
+        return false;
+    }
+
+    unitsMap units1Map;
+    updateUnitsMap(units1, units1Map);
```

Comment:
Urgh, this I do detest.  If I was doing this I would probably create a small shim function to create my units map and inside that function have these two lines of code.